### PR TITLE
Instrument MetricsReader with metrics

### PR DESCRIPTION
**Pull request by albertteoh**
## Which problem is this PR solving?
- Partially resolves: #3587

## Short description of the changes
- Adds metrics instrumentation around the MetricsReader.
- Uses the same pattern as https://github.com/jaegertracing/jaeger/blob/main/storage/spanstore/metrics/decorator.go.

## Testing

- Add unit tests.
- Manually tested both jaeger-query and all-in-one. Results below:

### jaeger-query

```
curl localhost:16687/metrics
```

#### Success

With Prometheus server running.

**requests**
```
jaeger_query_requests_total{operation="get_call_rates",result="err"} 0
jaeger_query_requests_total{operation="get_call_rates",result="ok"} 18
jaeger_query_requests_total{operation="get_error_rates",result="err"} 0
jaeger_query_requests_total{operation="get_error_rates",result="ok"} 18
jaeger_query_requests_total{operation="get_latencies",result="err"} 0
jaeger_query_requests_total{operation="get_latencies",result="ok"} 36
jaeger_query_requests_total{operation="get_min_step_duration",result="err"} 0
jaeger_query_requests_total{operation="get_min_step_duration",result="ok"} 0
```

**responses**
```
jaeger_query_responses_bucket{operation="get_call_rates",le="0.005"} 18
jaeger_query_responses_bucket{operation="get_error_rates",le="0.005"} 18
jaeger_query_responses_bucket{operation="get_latencies",le="0.005"} 36
```

**latency_buckets**
```
jaeger_query_latency_bucket{operation="get_call_rates",result="ok",le="0.005"} 5
jaeger_query_latency_bucket{operation="get_call_rates",result="ok",le="0.01"} 13
jaeger_query_latency_bucket{operation="get_call_rates",result="ok",le="0.025"} 18

jaeger_query_latency_bucket{operation="get_error_rates",result="ok",le="0.005"} 7
jaeger_query_latency_bucket{operation="get_error_rates",result="ok",le="0.01"} 13
jaeger_query_latency_bucket{operation="get_error_rates",result="ok",le="0.025"} 18

jaeger_query_latency_bucket{operation="get_latencies",result="ok",le="0.005"} 7
jaeger_query_latency_bucket{operation="get_latencies",result="ok",le="0.01"} 25
jaeger_query_latency_bucket{operation="get_latencies",result="ok",le="0.025"} 36
```

#### Fail

Prometheus server NOT running.

**requests**
```
jaeger_query_requests_total{operation="get_call_rates",result="err"} 4
jaeger_query_requests_total{operation="get_call_rates",result="ok"} 0
jaeger_query_requests_total{operation="get_error_rates",result="err"} 4
jaeger_query_requests_total{operation="get_error_rates",result="ok"} 0
jaeger_query_requests_total{operation="get_latencies",result="err"} 8
jaeger_query_requests_total{operation="get_latencies",result="ok"} 0
jaeger_query_requests_total{operation="get_min_step_duration",result="err"} 0
jaeger_query_requests_total{operation="get_min_step_duration",result="ok"} 0
```

### all-in-one

```
curl localhost:14269/metrics
```

#### Success

With Prometheus server running.

**requests**

```
jaeger_requests_total{operation="get_call_rates",result="err"} 0
jaeger_requests_total{operation="get_call_rates",result="ok"} 6
jaeger_requests_total{operation="get_error_rates",result="err"} 0
jaeger_requests_total{operation="get_error_rates",result="ok"} 6
jaeger_requests_total{operation="get_latencies",result="err"} 0
jaeger_requests_total{operation="get_latencies",result="ok"} 12
jaeger_requests_total{operation="get_min_step_duration",result="err"} 0
jaeger_requests_total{operation="get_min_step_duration",result="ok"} 0
```

**responses**
```
jaeger_responses_bucket{operation="get_call_rates",le="0.005"} 6
jaeger_responses_bucket{operation="get_error_rates",le="0.005"} 6
jaeger_responses_bucket{operation="get_latencies",le="0.005"} 12
jaeger_responses_bucket{operation="get_min_step_duration",le="0.005"} 0
```

**latency_buckets**
```
jaeger_latency_bucket{operation="get_call_rates",result="ok",le="0.005"} 0
jaeger_latency_bucket{operation="get_call_rates",result="ok",le="0.01"} 1
jaeger_latency_bucket{operation="get_call_rates",result="ok",le="0.025"} 4
jaeger_latency_bucket{operation="get_call_rates",result="ok",le="0.05"} 6

jaeger_latency_bucket{operation="get_error_rates",result="ok",le="0.005"} 0
jaeger_latency_bucket{operation="get_error_rates",result="ok",le="0.01"} 1
jaeger_latency_bucket{operation="get_error_rates",result="ok",le="0.025"} 4
jaeger_latency_bucket{operation="get_error_rates",result="ok",le="0.05"} 6

jaeger_latency_bucket{operation="get_latencies",result="ok",le="0.005"} 0
jaeger_latency_bucket{operation="get_latencies",result="ok",le="0.01"} 0
jaeger_latency_bucket{operation="get_latencies",result="ok",le="0.025"} 6
jaeger_latency_bucket{operation="get_latencies",result="ok",le="0.05"} 12
```

#### Fail

Prometheus server NOT running.

**requests**
```
jaeger_requests_total{operation="get_call_rates",result="err"} 6
jaeger_requests_total{operation="get_call_rates",result="ok"} 0
jaeger_requests_total{operation="get_error_rates",result="err"} 6
jaeger_requests_total{operation="get_error_rates",result="ok"} 0
jaeger_requests_total{operation="get_latencies",result="err"} 12
jaeger_requests_total{operation="get_latencies",result="ok"} 0
jaeger_requests_total{operation="get_min_step_duration",result="err"} 0
jaeger_requests_total{operation="get_min_step_duration",result="ok"} 0
```

### CODE REVIEW

File: cmd/all-in-one/main.go [hunk 1]
```diff
@@ -307,12 +308,23 @@ func initTracer(metricsFactory metrics.Factory, logger *zap.Logger) io.Closer {
 	return closer
 }
 
-func createMetricsQueryService(factory *metricsPlugin.Factory, v *viper.Viper, logger *zap.Logger) (querysvc.MetricsQueryService, error) {
-	if err := factory.Initialize(logger); err != nil {
+func createMetricsQueryService(
+	metricsReaderFactory *metricsPlugin.Factory,
+	metricsReaderMetricsFactory metrics.Factory,
```
Comment (yurishkuro) [inline cmd/all-in-one/main.go:313]: Group logger and metrics next to each other

File: storage/metricsstore/metrics/decorator.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+	getLatenciesMetrics       *queryMetrics
+	getCallRatesMetrics       *queryMetrics
+	getErrorRatesMetrics      *queryMetrics
+	getMinStepDurationMetrics *queryMetrics
+}
+
+type queryMetrics struct {
+	Errors     metrics.Counter `metric:"requests" tags:"result=err"`
+	Successes  metrics.Counter `metric:"requests" tags:"result=ok"`
+	Responses  metrics.Timer   `metric:"responses"` // used as a histogram, not necessary for GetTrace
```
Comment (yurishkuro) [inline storage/metricsstore/metrics/decorator.go:39]: Copied comment?

File: storage/metricsstore/metrics/decorator.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+}
+
+func (q *queryMetrics) emit(err error, latency time.Duration, responses int) {
+	if err != nil {
+		q.Errors.Inc(1)
+		q.ErrLatency.Record(latency)
+	} else {
+		q.Successes.Inc(1)
+		q.OKLatency.Record(latency)
+		q.Responses.Record(time.Duration(responses))
```
Comment (yurishkuro) [inline storage/metricsstore/metrics/decorator.go:51]: I don't follow, what is 1ns here?

File: cmd/all-in-one/main.go (hunk 1 above)
Comment (albertteoh) [reply cmd/all-in-one/main.go]: Addressed in [a41b9be](https://github.com/jaegertracing/jaeger/pull/3667/commits/a41b9bebc8fd6932cadd56c86fb9d97229c591f3).

File: storage/metricsstore/metrics/decorator.go (hunk 3 above)
Comment (albertteoh) [reply storage/metricsstore/metrics/decorator.go]: Good question, I actually don't really understand how it was used in [spanstore](https://github.com/jaegertracing/jaeger/blob/bf64373d1e690594fd8c279720faf32722cf1494/storage/spanstore/metrics/decorator.go#L53) either. 

Either way, I don't think it applies for the metrics reader use case so I've removed it in [857f64a](https://github.com/jaegertracing/jaeger/pull/3667/commits/857f64a8c066fa0ec8c8d7532c56212122da35a4).

File: storage/metricsstore/metrics/decorator.go (hunk 2 above)
Comment (albertteoh) [reply storage/metricsstore/metrics/decorator.go]: Removed in [857f64a](https://github.com/jaegertracing/jaeger/pull/3667/commits/857f64a8c066fa0ec8c8d7532c56212122da35a4).

**Review by yurishkuro — APPROVED** [review]

**Pull request merged**